src/ripple/app/main/Application.cpp

```diff
@@ -1110,7 +1064,30 @@ class ApplicationImp : public Application, public RootStoppable, public BasicApp
                 return validators().trustedPublisher(pubKey);
             });
 
-        stopped();
+        // The order of these stop calls is delicate.
+        // Re-ordering them risks undefined behavior.
+        m_shaMapStore->stop();
+        m_jobQueue->stop();
+        if (shardArchiveHandler_)
+            shardArchiveHandler_->stop();
+        m_nodeStore->stop();
+        if (shardStore_)
+            shardStore_->stop();
+        if (config_->reporting())
+        {
+#ifdef RIPPLED_REPORTING
+            pgPool_->stop();
+#endif
+            reportingETL_->stop();
+        }
+        m_inboundLedgers->stop();
+        m_inboundTransactions->stop();
+        overlay_->stop();
+        perfLog_->stop();
+        grpcServer_->stop();
+        m_networkOPs->stop();
+        m_ledgerMaster->stop();
+        m_loadManager->stop();
```

Comment:
`The order of these stop calls is delicate`. I am concerned about this. I feel that this situation is exactly the problem that `Stoppable` was meant to solve (managing the order in which things are stopped); with this PR, now we have to manually maintain this order.

Reply:
We had to manually maintain the order before. Now the order is visible. That's an improvement.

Reply:
Should we maintain a stack in `ApplicationImpl`, that we can simply `pop` from to get things in the right order?

Reply:
I dug through the remaining stoppable objects (everything in this block that has its `stop` method called) to try to figure out the dependency graph.

1. Each object is a singleton of its type, so we can represent them by their types. The exceptions are both `NodeStore::Database`, but the second database is always a `NodeStore::DatabaseShard`, so we can distinguish them with those types.
2. Objects tend to depend on each other by going through a getter method in `Application`.
3. Two classes, `NetworkOPs` and `LedgerMaster` are special. They depend on many objects that depend on them, creating cycles, but they are treated like low-level abstractions. The separation between `Application` and `ApplicationImp` is what lets us tie the recursive knot.
4. `JobQueue` is very low level in the hierarchy, but it stops early because some other objects only stop when they cannot schedule any more jobs. Objects that depend on `JobQueue` will not break if it stops before they do.

In the hierarchy below, with the exception of `JobQueue` and `NetworkOPs`, no object depends on another object at the same or higher level... to my best belief. Some stoppables depend on other stoppables indirectly, and while I tried to chase all those connections down, there are many classes and LOC involved here; I may have missed something.

Level 0:
- `PerfLog`
- `NetworkOPs`
- `LedgerMaster`
- `NodeStore::Database`
- `PgPool`

Level 1:
- `JobQueue`
- `SHAMapStore`
- `Overlay`

Level 2:
- `LoadManager`
- `GRPCServer`
- `InboundTransactions`
- `NodeStore::DatabaseShard`

Level 3:
- `RPC::ShardArchiveHandler`
- `InboundLedgers`

Level 4:
- `LedgerReplayer`
- `LedgerCleaner`

Reply:
Further, `LedgerMaster` is only a "stoppable" because it owns `LedgerCleaner`. The connection between them is very weak, and I plan to move `LedgerCleaner` to be a member of `Application` directly. No stoppable depends on `LedgerCleaner`.